plugin/storage/badger/spanstore/reader.go

```diff
@@ -0,0 +1,488 @@
+
+	errNoTraces = errors.New("No trace with that ID found")
+
+	defaultMaxDuration = model.DurationAsMicroseconds(time.Hour * 24)
+)
+
+const (
+	defaultNumTraces = 100
+	sizeOfTraceID    = 16
+)
+
+// TraceReader reads traces from the local badger store
+type TraceReader struct {
+	store *badger.DB
+	cache *CacheStore
+}
+
+// NewTraceReader returns a TraceReader with cache
+func NewTraceReader(db *badger.DB, c *CacheStore) *TraceReader {
+	return &TraceReader{
+		store: db,
+		cache: c,
+	}
+}
+
+// getTraces enriches TraceIDs to Traces
+func (r *TraceReader) getTraces(traceIDs []model.TraceID) ([]*model.Trace, error) {
+	// Get by PK
+	traces := make([]*model.Trace, 0, len(traceIDs))
+	prefixes := make([][]byte, 0, len(traceIDs))
+
+	for _, traceID := range traceIDs {
+		prefixes = append(prefixes, createPrimaryKeySeekPrefix(traceID))
+	}
+
+	err := r.store.View(func(txn *badger.Txn) error {
+		opts := badger.DefaultIteratorOptions
+		opts.PrefetchSize = 10 // TraceIDs are not sorted, pointless to prefetch large amount of values
+		it := txn.NewIterator(opts)
+		defer it.Close()
+
+		val := []byte{}
+		for _, prefix := range prefixes {
+			spans := make([]*model.Span, 0, 4) // reduce reallocation requirements by defining some initial length
+
+			for it.Seek(prefix); it.ValidForPrefix(prefix); it.Next() {
+				// Add value to the span store (decode from JSON / defined encoding first)
+				// These are in the correct order because of the sorted nature
+				item := it.Item()
+				val, err := item.ValueCopy(val)
+				if err != nil {
+					return err
+				}
+
+				sp := model.Span{}
+				switch item.UserMeta() & 0x0F {
+				case jsonEncoding:
+					if err := json.Unmarshal(val, &sp); err != nil {
+						return err
+					}
+				case protoEncoding:
+					if err := proto.Unmarshal(val, &sp); err != nil {
+						return err
+					}
+				default:
+					return fmt.Errorf("Unknown encoding type: %04b", item.UserMeta()&0x0F)
+				}
+				spans = append(spans, &sp)
+			}
+			trace := &model.Trace{
+				Spans: spans,
+			}
+			traces = append(traces, trace)
+		}
+		return nil
+	})
+
+	// TODO Do the Unmarshal here so we can release the transaction earlier (we would pay with extra allocations..)
+
+	return traces, err
+
+}
+
+// GetTrace takes a traceID and returns a Trace associated with that traceID
+func (r *TraceReader) GetTrace(traceID model.TraceID) (*model.Trace, error) {
+	traces, err := r.getTraces([]model.TraceID{traceID})
+	if err != nil {
+		return nil, err
+	} else if len(traces) == 1 {
+		return traces[0], nil
+	}
+
+	return nil, nil
+}
+
+func createPrimaryKeySeekPrefix(traceID model.TraceID) []byte {
+	buf := new(bytes.Buffer)
+	buf.WriteByte(spanKeyPrefix)
+	binary.Write(buf, binary.BigEndian, traceID.High)
+	binary.Write(buf, binary.BigEndian, traceID.Low)
+	return buf.Bytes()
+}
+
+// GetServices fetches the sorted service list that have not expired
+func (r *TraceReader) GetServices() ([]string, error) {
+	return r.cache.GetServices()
```

Comment:
I noticed an issue with this implementation that in case of initializing badger factory by reloading from a pre-existing directory, the services list does not get populated with services from pre-existing traces even though the traces are accessible by making a direct REST API call.
Probably the cause is that you are relying on the cache which is not pre-populated.